### PR TITLE
 added comments to EventEmitter and more ...

### PR DESCRIPTION
**Pull request by jefleponot**
This feature adds only hooks to filter events : 
it allows callbacks for "page.confirm" and "page.prompt"...
So everyOne can use library to render dialogue Box

```
casper.on("page.confirm", function confirmRender(msg, ret){ 
    self.renderDialogBox(msg, null, true);
});
casper.on("page.prompt", function promptRender(msg, value, ret) {
    self.renderDialogBox(msg, value, true);
});
casper.on("remote.alert", function alertRender(msg) {
    self.renderDialogBox(msg, null, false);
});
```

adds also comments for events.js functions

### CODE REVIEW

File: modules/events.js [hunk 1]
```diff
@@ -32,286 +32,403 @@
...
+    "use strict";
+    var type = arguments[0], res;
+    if (!this._filters) {
+        this._filters = {};
+    }
+    var _filter = this._filters[type];
+    if (typeof _filter === 'function') {
+        res = _filter.apply(this, Array.prototype.slice.call(arguments, 1));
+    }
+    this.emit(Array.prototype.slice.call(arguments, 0).concat([res]));
```
Comment (jefleponot) [inline modules/events.js]: this only line has been added :  
   this.emit(Array.prototype.slice.call(arguments, 0).concat([res]));

**Comment by paazmaya** [comment]
Thank you for the syntax changes, but for the possible added functionality I would like to hear others opinions before merging

**Comment by jefleponot** [comment]
Hi @paazmaya , thanks to come back to life one of my old PR.

File: modules/events.js [hunk 2]
```diff
@@ -32,286 +32,403 @@
...
+exports.EventEmitter = EventEmitter;
 var defaultMaxListeners = 10;
+
+/**
+ * By default EventEmitters will print a warning if more than 10 listeners
+ * are added for a particular event. This is a useful default that helps finding memory leaks.
+ * Obviously, not all events should be limited to just 10 listeners.
+ * The emitter.setMaxListeners() method allows the limit to be modified for
+ * this specific EventEmitter instance. The value can be set to Infinity (or 0)
+ * for to indicate an unlimited number of listeners.
```
Comment (istr) [inline modules/events.js]: ~~for~~

File: modules/events.js [hunk 3]
```diff
@@ -32,286 +32,403 @@
...
+ *
+ * @param  Integer  nb  Max Number of Listeners
```
Comment (BIGjuevos) [inline modules/events.js]: the variable in the docs docs does not match the var in the function definition

**Comment by jefleponot** [comment]
@istr , @BIGjuevos , @paazmaya , could we merge it now ?

File: modules/events.js [hunk 4]
```diff
@@ -32,286 +32,403 @@
...
+    if (!this._events[type]) {
+        // Optimize the case of one listener. Don't need the extra array object.
+        this._events[type] = listener;
+    } else if (isArray(this._events[type])) {
+        // If we've already got an array, just append.
+        this._events[type][ type === 'fail' ? 'unshift' : 'push'](listener);
+
+        // Check for listener leak
+        if (!this._events[type].warned) {
+            var m;
```
Comment (mickaelandrieu) [inline modules/events.js]: can you use a more explicit variable name ? Are we ok this is a kind a stack of listeners atm ?

File: modules/events.js [hunk 5]
```diff
@@ -32,286 +32,403 @@
...
+            if (typeof this._maxListeners !== 'undefined') {
+                m = this._maxListeners;
+            } else {
+                m = defaultMaxListeners;
+            }
+
+            if (m && m > 0 && this._events[type].length > m) {
```
Comment (mickaelandrieu) [inline modules/events.js]: it may be logic, but I dont like a repetition of code, can we improve it ?

**Comment by mickaelandrieu** [comment]
Sorry for this (late) review, but I dont understand well what is done in this pull request, so I can't realy vote for or against this ^^

**Comment by jefleponot** [comment]
Hi @mickaelandrieu,

1. The module file "events.js" is the only one that is without comment.
To my mind, it looks like a copy paste from https://github.com/asyncly/EventEmitter2/blob/master/lib/eventemitter2.js or its forefather. (with 2 functions added)

2. This PR considers that filters can also trigger events. (1 line : have a look above)

3. To conclude, I tried to reduce eslint warnings when it was doable.

Regards

**Comment by jefleponot** [comment]
@mickaelandrieu , please merge or close this feature

File: modules/events.js [hunk 6]
```diff
@@ -32,286 +32,403 @@
...
+        this._filters = {};
+    }
+    if (typeof filterFn !== 'function') {
+        throw new CasperError('setFilter only takes instances of Function');
+    }
+    if (!this._filters[type]) {
+        this._filters[type] = filterFn;
+        return true;
+    }
+    // TODO: process multiple filters? in which order? disallow?
```
Comment (mickaelandrieu) [inline modules/events.js]: how about this todo ?

@casperjs/contributors would you mind to take a look ? :+1:

**Comment by jefleponot** [comment]
@istr , @BIGjuevos any opinion ?

**Comment by jefleponot** [comment]
Re

**Pull request merged**